demos/coreMQTT_Agent/mqtt_agent_task.c

```diff
@@ -0,0 +1,1025 @@
+ * prvMQTTAgentTask() to become the agent task itself.
+ *
+ * This task calls MQTTAgent_CommandLoop() in a loop, until MQTTAgent_Terminate()
+ * is called. If an error occurs in the command loop, then it will reconnect the
+ * TCP and MQTT connections.
+ *
+ * @param[in] pvParameters Parameters as passed at the time of task creation. Not
+ * used in this example.
+ */
+static void prvMQTTAgentTask( void * pvParameters );
+
+/**
+ * @brief The main task used in the MQTT demo.
+ *
+ * This task creates the network connection and all other demo tasks. Then,
+ * rather than create prvMQTTAgentTask() as a separate task, it simply calls
+ * prvMQTTAgentTask() to become the agent task itself.
+ *
+ * @param[in] pvParameters Parameters as passed at the time of task creation. Not
+ * used in this example.
+ *
+ * @return EXIT_SUCCESS if demo completes successfully, else EXIT_FAILURE.
+ */
+static int prvConnectAndCreateDemoTasks( void * pvParameters );
+
+/**
+ * @brief The timer query function provided to the MQTT context.
+ *
+ * @return Time in milliseconds.
+ */
+static uint32_t prvGetTimeMs( void );
+
+/**
+ * @brief Connects a TCP socket to the MQTT broker, then creates and MQTT
+ * connection to the same.
+ */
+static BaseType_t prvConnectToMQTTBroker( void );
+
+/*
+ * Functions that start the tasks demonstrated by this project.
+ */
+
+extern void vStartSimpleSubscribePublishTask( uint32_t ulTaskNumber,
+                                              configSTACK_DEPTH_TYPE uxStackSize,
+                                              UBaseType_t uxPriority,
+                                              struct DemoParams * pxParams );
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The network context used by the MQTT library transport interface.
+ * See https://www.freertos.org/network-interface.html
+ */
+static NetworkContext_t xNetworkContext;
+
+/**
+ * @brief The parameters for the network context using a TLS channel.
+ */
+static SecureSocketsTransportParams_t secureSocketsTransportParams;
+
+/**
+ * @brief Global entry time into the application to use as a reference timestamp
+ * in the #prvGetTimeMs function. #prvGetTimeMs will always return the difference
+ * between the current time and the global entry time. This will reduce the chances
+ * of overflow for the 32 bit unsigned integer used for holding the timestamp.
+ */
+static uint32_t ulGlobalEntryTimeMs;
+
+MQTTAgentContext_t xGlobalMqttAgentContext;
+
+static uint8_t xNetworkBuffer[ MQTT_AGENT_NETWORK_BUFFER_SIZE ];
+
+static MQTTAgentMessageContext_t xCommandQueue;
+
+static struct DemoParams taskParameters[ democonfigNUM_SIMPLE_SUB_PUB_TASKS_TO_CREATE ];
+
+/**
+ * @brief The global array of subscription elements.
+ *
+ * @note No thread safety is required to this array, since the updates the array
+ * elements are done only from one task at a time. The subscription manager
```

Comment:
```suggestion
 * @note No thread safety is required to this array, since updates to the array
 * elements are done only from one task at a time. The subscription manager
```